### PR TITLE
QL: case sensitive support in EQL

### LINKED ISSUE #55340: QL: Optimize startWith to prefix queries where possible
When dealing with literals, the `startsWith` function can be translated to a [prefix query](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-prefix-query.html) and fallback to a script only when dealing with a scalar.

### PR DESCRIPTION
**Pull request by astefan**
This PR adds several enhancements:
* adds a generic startsWith function to QL
* modifies the existent EQL startsWith function to be case sensitive aware
* improves the existent EQL startsWith function to use a `prefix` query when the function is used in a case sensitive context. Same improvement is used in SQL's newly added STARTS_WITH function.
* adds case sensitivity to EQL configuration through a `case_sensitive` parameter in the eql request, as established in https://github.com/elastic/elasticsearch/issues/54411. The `case_sensitive` parameter can be specified when running queries (default is case insensitive)

What this PR is not covering:
* case sensitivity in other string related EQL functions
* case sensitivity aware integration testing in EQL

Fixes https://github.com/elastic/elasticsearch/issues/55340.
Relates to https://github.com/elastic/elasticsearch/issues/54411.

### CODE REVIEW

File: docs/reference/sql/functions/string.asciidoc [hunk 1]
```diff
@@ -422,6 +422,30 @@ SPACE(count) <1>
...
+    pattern)  <2>
+--------------------------------------------------
+*Input*:
+
+<1> string expression
+<2> string expression
+
+*Output*: boolean value
+
+*Description*: Returns `true` if the source expression starts with the specified pattern, `false` otherwise. The matching is case sensitive.
```
Comment (matriv) [inline docs/reference/sql/functions/string.asciidoc:442]: maybe add info about the null handling?

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/expression/function/scalar/string/StartsWith.java [hunk 2]
```diff
@@ -6,115 +6,20 @@
...
-
-    @Override
-    public Expression replaceChildren(List<Expression> newChildren) {
-        if (newChildren.size() != 2) {
-            throw new IllegalArgumentException("expected [2] children but received [" + newChildren.size() + "]");
-        }
-
-        return new StartsWith(source(), newChildren.get(0), newChildren.get(1));
+    public boolean isCaseSensitive() {
+        return ((EqlConfiguration) configuration()).isCaseSensitive();
```
Comment (matriv) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/expression/function/scalar/string/StartsWith.java:22]: Maybe add isCaseSensitive to the ql configuration with default value and avoid the cast?

File: x-pack/plugin/eql/src/main/resources/org/elasticsearch/xpack/eql/plugin/eql_whitelist.txt [hunk 3]
```diff
@@ -18,6 +18,11 @@ class org.elasticsearch.xpack.ql.expression.function.scalar.whitelist.InternalQl
   double nullSafeSortNumeric(Number)
   String nullSafeSortString(Object)
 
+#
```
Comment (matriv) [inline x-pack/plugin/eql/src/main/resources/org/elasticsearch/xpack/eql/plugin/eql_whitelist.txt:21]: Why did you add an extra ASCII Functions block?
https://github.com/elastic/elasticsearch/pull/56404/files#diff-610841f585d6bbff1fd42fb91c3c424aR65

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/analysis/VerifierTests.java [hunk 4]
```diff
@@ -1,11 +1,12 @@
 /*
- * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+  * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
```
Comment (matriv) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/analysis/VerifierTests.java:2]: minor: whitespace

File: x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/expression/function/scalar/string/StartsWithFunctionPipeTests.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+                f.expression(),
+                f.field(),
+                pipe(((Expression) randomValueOtherThan(f.pattern(), () -> randomStringLiteral()))),
+                f.isCaseSensitive()));
+        randoms.add(f -> new StartsWithFunctionPipe(f.source(),
+                f.expression(),
+                pipe(((Expression) randomValueOtherThan(f.field(), () -> randomStringLiteral()))),
+                pipe(((Expression) randomValueOtherThan(f.pattern(), () -> randomStringLiteral()))),
+                f.isCaseSensitive()));
+
```
Comment (matriv) [inline x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/expression/function/scalar/string/StartsWithFunctionPipeTests.java:115]: Could you please add test with just changing the case sensitive flag?
and maybe one more where all of field/pattern/caseSensitive/flag are changed.

File: x-pack/plugin/sql/src/main/resources/org/elasticsearch/xpack/sql/plugin/sql_whitelist.txt [hunk 6]
```diff
@@ -29,6 +29,11 @@ class org.elasticsearch.xpack.ql.expression.function.scalar.whitelist.InternalQl
   double nullSafeSortNumeric(Number)
   String nullSafeSortString(Object)
 
+#
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/resources/org/elasticsearch/xpack/sql/plugin/sql_whitelist.txt:32]: Same here, there is a block for ascii functions: https://github.com/elastic/elasticsearch/pull/56404/files#diff-edef662ace9df6873edfc33572672eeeR144

File: x-pack/plugin/sql/qa/src/main/resources/docs/docs.csv-spec [hunk 7]
```diff
@@ -1798,6 +1799,16 @@ SELECT SPACE(3);
...
+stringStartsWith
+// tag::stringStartsWith
+SELECT STARTS_WITH('Elasticsearch', 'Elastic');
+
+STARTS_WITH('Elasticsearch', 'Elastic')
+--------------------------------
+true    
+// end::stringStartsWith
+;
+
```
Comment (matriv) [inline x-pack/plugin/sql/qa/src/main/resources/docs/docs.csv-spec:1811]: Maybe and example with different case so it returns false.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/scalar/string/StartsWithFunctionProcessor.java [hunk 8]
```diff
@@ -20,43 +20,51 @@
 
     private final Processor source;
     private final Processor pattern;
+    private final boolean isCaseSensitive;
 
-    public StartsWithFunctionProcessor(Processor source, Processor pattern) {
+    public StartsWithFunctionProcessor(Processor source, Processor pattern, boolean isCaseSensitive) {
```
Comment (matriv) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/scalar/string/StartsWithFunctionProcessor.java:25]: Maybe remove the `Function` as we don't normally use it in processor names.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/scalar/string/StartsWithFunctionPipe.java [hunk 9]
```diff
@@ -17,13 +17,15 @@
 
 public class StartsWithFunctionPipe extends Pipe {
```
Comment (matriv) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/scalar/string/StartsWithFunctionPipe.java:18]: Maybe remove `Function` from the name.

**Review by matriv — COMMENTED** [review]
Looks good, left some comments mostly minor.

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/planner/QueryFolderOkTests.java [hunk 10]
```diff
@@ -114,6 +110,11 @@ private static void addSpec(ArrayList<Object[]> arr, String name, String query,
     }
 
     public void test() {
+        // skip tests that do not make sense from case sensitivity point of view
```
Comment (costin) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/planner/QueryFolderOkTests.java:113]: It's idiomatic to use `assume` which will indicate to the framework that the test has not actually ran.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/PlanExecutor.java [hunk 11]
```diff
@@ -22,7 +22,7 @@
 import org.elasticsearch.xpack.sql.planner.Planner;
 import org.elasticsearch.xpack.sql.planner.PlanningException;
 import org.elasticsearch.xpack.sql.proto.SqlTypedParamValue;
-import org.elasticsearch.xpack.sql.session.Configuration;
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/PlanExecutor.java:25]: What's the advantage of using `Configuration` instead of `SqlConfiguration`?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/NoArgumentsConfigurationFunction.java [hunk 12]
```diff
@@ -17,14 +17,14 @@
 import java.util.List;
 import java.util.Objects;
 
-public abstract class ConfigurationFunction extends ScalarFunction {
+import static java.util.Collections.emptyList;
+
+public abstract class NoArgumentsConfigurationFunction extends ConfigurationFunction {
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/NoArgumentsConfigurationFunction.java:22]: The name is unfortunate - how about `SqlConfigurationFunction`, `BasicConfigurationFunction` or `DefaultConfigurationFunction`, etc...

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/session/EqlSession.java [hunk 13]
```diff
@@ -27,22 +29,22 @@
...
-            Optimizer optimizer, Planner planner, PlanExecutor planExecutor) {
+    public EqlSession(Client client, EqlConfiguration cfg, IndexResolver indexResolver, PreAnalyzer preAnalyzer,
+            FunctionRegistry functionRegistry, Optimizer optimizer, Planner planner, PlanExecutor planExecutor) {
 
         this.client = new ParentTaskAssigningClient(client, cfg.getTaskId());
         this.configuration = cfg;
         this.indexResolver = indexResolver;
         this.preAnalyzer = preAnalyzer;
-        this.analyzer = analyzer;
+        this.analyzer = new Analyzer(cfg, functionRegistry, new Verifier());
```
Comment (costin) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/session/EqlSession.java:47]: That's unfortunate - do we need a Verifier instance each time?

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/expression/function/scalar/string/StartsWith.java (hunk 2 above)
Comment (costin) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/expression/function/scalar/string/StartsWith.java:22]: I disagree - that would cause `SqlConfiguration` to have a case sensitivity option that is meaningless.

**Review by costin — APPROVED** [review]
Left a few comments but looks good to me otherwise.
Thanks for the various cleanups.

File: x-pack/plugin/eql/src/main/resources/org/elasticsearch/xpack/eql/plugin/eql_whitelist.txt (hunk 3 above)
Comment (astefan) [reply x-pack/plugin/eql/src/main/resources/org/elasticsearch/xpack/eql/plugin/eql_whitelist.txt:21]: This ASCII block is under a different ScriptUtils class. The method comes via QL, whereas the other ASCII block handles EQL specific functions.

Comment (matriv) [reply x-pack/plugin/eql/src/main/resources/org/elasticsearch/xpack/eql/plugin/eql_whitelist.txt:21]: Sorry, missed that.

File: x-pack/plugin/sql/src/main/resources/org/elasticsearch/xpack/sql/plugin/sql_whitelist.txt (hunk 6 above)
Comment (astefan) [reply x-pack/plugin/sql/src/main/resources/org/elasticsearch/xpack/sql/plugin/sql_whitelist.txt:32]: Yep, same here. Different ScriptUtils class that feeds the painless method in sql.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/scalar/string/StartsWithFunctionProcessor.java (hunk 8 above)
Comment (astefan) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/scalar/string/StartsWithFunctionProcessor.java:25]: There are functions where this pattern is used - String functions.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/PlanExecutor.java (hunk 11 above)
Comment (astefan) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/PlanExecutor.java:25]: This is just a rename. The SQL's Configuration => SqlConfiguration.
The same for Eql. EQL's Configuration was renamed to EqlConfiguration.

I found it a bit confusing to have Configuration in three separate projects.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/NoArgumentsConfigurationFunction.java (hunk 12 above)
Comment (astefan) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/NoArgumentsConfigurationFunction.java:22]: Agree. Will change.

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/session/EqlSession.java (hunk 13 above)
Comment (astefan) [reply x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/session/EqlSession.java:47]: You are correct. We can create the Verifier _once_ in the plan executor and pass that along.

File: x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/expression/function/scalar/string/StartsWithFunctionPipeTests.java [hunk 14]
```diff
@@ -97,21 +99,18 @@ public void testReplaceChildren() {
...
-                f.expression(),
-                f.field(),
-                pipe(((Expression) randomValueOtherThan(f.pattern(), () -> randomStringLiteral()))),
-                f.isCaseSensitive()));
-        randoms.add(f -> new StartsWithFunctionPipe(f.source(),
-                f.expression(),
-                pipe(((Expression) randomValueOtherThan(f.field(), () -> randomStringLiteral()))),
-                pipe(((Expression) randomValueOtherThan(f.pattern(), () -> randomStringLiteral()))),
-                f.isCaseSensitive()));
+        for(int i = 1; i < 4; i++) {
```
Comment (matriv) [inline x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/expression/function/scalar/string/StartsWithFunctionPipeTests.java:102]:
```suggestion
        for (int i = 1; i < 4; i++) {
```

File: x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/expression/function/scalar/string/StartsWithFunctionPipeTests.java [hunk 15]
```diff
@@ -97,21 +99,18 @@ public void testReplaceChildren() {
...
+            for(BitSet comb : new Combinations(3, i)) {
```
Comment (matriv) [inline x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/expression/function/scalar/string/StartsWithFunctionPipeTests.java:103]:
```suggestion
            for (BitSet comb : new Combinations(3, i)) {
```

**Review by matriv — APPROVED** [review]
LGTM. left 2 minor formatting comments.
